bigquery/storage/managedwriter/managed_stream.go

```diff
@@ -140,3 +159,173 @@ func (ms *ManagedStream) Finalize(ctx context.Context) (int64, error) {
+	return ms.arc, ms.pending, ms.err
+}
+
+// openWithRetry is responsible for navigating the (re)opening of the underlying stream connection.
+// The expectation is this is only ever invoked when the caller has the mutex lock.
+func (ms *ManagedStream) openWithRetry() (storagepb.BigQueryWrite_AppendRowsClient, chan *pendingWrite, error) {
+	r := defaultRetryer{}
+	for {
+		arc, err := ms.open()
+		bo, shouldRetry := r.Retry(err)
+		if err != nil && shouldRetry {
+			if err := gax.Sleep(ms.ctx, bo); err != nil {
+				return nil, nil, err
+			}
+			continue
+		}
+		if err == nil {
+			// The channel relationship with its ARC is 1:1.  If we get a new ARC, create a new chan
+			// and fire up the associated receive processor.
+			ch := make(chan *pendingWrite)
+			go recvProcessor(ms.ctx, arc, ch)
+			return arc, ch, nil
+		}
+		return arc, nil, err
+	}
+}
+
+// call serves as a closure that forwards the call to a (possibly reopened) AppendRowsClient), and the associated
+// pendingWrite channel for the ARC connection.
+func (ms *ManagedStream) call(f func(storagepb.BigQueryWrite_AppendRowsClient, chan *pendingWrite) error, opts ...gax.CallOption) error {
+	var settings gax.CallSettings
+	for _, opt := range opts {
+		opt.Resolve(&settings)
+	}
+	var r gax.Retryer = &defaultRetryer{}
+	if settings.Retry != nil {
+		r = settings.Retry()
+	}
+
+	var arc *storagepb.BigQueryWrite_AppendRowsClient
+	var ch chan *pendingWrite
+	var err error
+
+	for {
+		arc, ch, err = ms.getStream(arc)
+		if err != nil {
+			return err
+		}
+		err = f(*arc, ch)
+		if err != nil {
+			bo, shouldRetry := r.Retry(err)
+			if shouldRetry {
+				if err := gax.Sleep(ms.ctx, bo); err != nil {
+					return err
+				}
+				continue
+			}
+			ms.mu.Lock()
+			ms.err = err
+			ms.mu.Unlock()
+		}
+		return err
+	}
+}
+
+func (ms *ManagedStream) append(pw *pendingWrite) error {
+	return ms.call(func(arc storagepb.BigQueryWrite_AppendRowsClient, ch chan *pendingWrite) error {
+		// TODO: we should only send stream ID and schema for the first message in a new stream, but
+		// we've decoupled the individual appends from caring about state of the connection.  We likely
+		// need to do this via a mutex-guarded state flag.
+		pw.request.WriteStream = ms.streamSettings.streamID
+		pw.request.GetProtoRows().WriterSchema = &storagepb.ProtoSchema{
+			ProtoDescriptor: ms.schemaDescriptor,
+		}
+		err := arc.Send(pw.request)
+		if err == nil {
+			ch <- pw
+		}
+		return err
+	})
+}
+
+func (ms *ManagedStream) CloseSend() error {
+	err := ms.call(func(arc storagepb.BigQueryWrite_AppendRowsClient, ch chan *pendingWrite) error {
+		err := arc.CloseSend()
+		if err == nil {
+			close(ch)
+		}
+		return err
+	})
+	ms.mu.Lock()
+	ms.err = io.EOF
+	close(ms.pending)
+	ms.mu.Unlock()
+	return err
+}
+
+// AppendRows sends the append requests to the service, and returns one AppendResult per row.
+func (ms *ManagedStream) AppendRows(data [][]byte, offset int64) ([]*AppendResult, error) {
```

Comment:
We should eventually add godoc examples for some of these.

Reply:
Yeah, there will need to be extensive docs on this thing, even with the simplification.